vexcl/pointer_qualifiers.hpp

```diff
@@ -0,0 +1,121 @@
+#ifndef VEXCL_POINTER_QUALIFIERS_HPP
+#define VEXCL_POINTER_QUALIFIERS_HPP
+
+
+#include <vexcl/operations.hpp>
+/// \cond INTERNAL
+namespace vex {
+	//_________________________________________________ constant keyword
+	// The struct will be used as an identificator of constant vector
+	// expressions.
+	struct constant_vector_terminal {};
+
+	// Introduce the new kind of expression to Boost.Proto.
+	typedef vector_expression<
+	typename boost::proto::terminal< constant_vector_terminal >::type
+	> constant_vector_expression;
+
+	// The actual terminal.
+	// This struct will hold a reference to a vector that needs to be declared with
+	// 'restrict' qualifier.
+	template <class T>
+	struct constant_vector : constant_vector_expression {
+		typedef T value_type; // has to be here for vex::deduce() to work with constant vectors.
+		const vector<T> &v;
+		constant_vector(const vector<T> &v) : v(v) {}
+	};
+
+
+
+	// Code generation helpers.
+	namespace traits {
+
+		// Tell proto it is ok to use these terminals in vexcl expressions.
+		template <>
+		struct is_vector_expr_terminal< constant_vector_terminal > : std::true_type {};
+
+		// Tell proto that it is not required to unpack the terminal with
+		// proto::value() function.
+		template <>
+		struct proto_terminal_is_value< constant_vector_terminal > : std::true_type {};
+
+		// Declare kernel parameter for the constant vector.
+		// This is the only specification that is different from vector<T>.
+		template <typename T>
+		struct kernel_param_declaration< constant_vector<T> > {
+			static void get(backend::source_generator &src,
+							const constant_vector<T>&,
+							const backend::command_queue&, const std::string &prm_name,
+							detail::kernel_generator_state_ptr)
+			{
+				// since 'restrict' keyword should go in between type and parameter
+				// name, we can just prepend parameter name with 'restrict '.
+#if defined(VEXCL_BACKEND_CUDA)
+				src.parameter< constant_ptr<T> >( prm_name);
+#else // defined (VEXCL_BACKEND_OPENCL)
+				src.parameter< constant_ptr<T> >( prm_name);
+#endif
```

Comment:
yea i was not sure if it needed a restricted keyword or not, but then i searched it up and it wasn't needed.